### PR TITLE
[Storage] Use thread pools to make IO operations async.

### PR DESCRIPTION
**Pull request by grao1991**
### Description

This can 10x state-sync TPS when these read hit disk.

### Test Plan
Tested on Alden-net.

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/3502)

### CODE REVIEW

File: storage/storage-interface/src/cached_state_view.rs [hunk 1]
```diff
@@ -104,13 +111,20 @@ impl CachedStateView {
...
 
     pub fn prime_cache_by_write_set(&self, write_sets: &[WriteSet]) -> Result<()> {
-        write_sets
-            .iter()
-            .flat_map(|write_set| write_set.iter())
-            .map(|(key, _)| key)
-            .collect::<HashSet<_>>()
-            .into_iter()
-            .try_for_each(|key| self.get_state_value(key).map(|_| ()))
+        IO_POOL.scope(|s| {
```
Comment (msmouse) [inline storage/storage-interface/src/cached_state_view.rs:114]: 1. why this not par_iter()?
2. this uses the SyncProofFetcher, does it matter? (should not)

Comment (grao1991) [reply storage/storage-interface/src/cached_state_view.rs:114]: 1. It's IO bound.. so the bottleneck will be the get_state_value itself.
2. no

**Review by msmouse — APPROVED** [review]

**Review by zekun000 — COMMENTED** [review]
is it worth spawning every key into a task? or batch multiple instead?

**Comment by grao1991** [comment]
> is it worth spawning every key into a task? or batch multiple instead?

In my understanding, the overhead is small because we use threadpool. Picking a thread is much faster than doing IO. Also the variance of latency among IO requests can be large, if slow requests are accidentally batched together the cost will be huge.

File: execution/executor/src/components/chunk_output.rs [hunk 2]
```diff
@@ -50,17 +48,14 @@ impl ChunkOutput {
...
         // collect all accounts touched and dedup
-        let state_updates = transaction_outputs
-            .iter()
-            .flat_map(|o| o.write_set())
-            .collect::<HashSet<_>>();
+        let write_set = transaction_outputs
+            .clone()
+            .into_iter()
+            .map(|o| o.into().0)
+            .collect::<Vec<_>>();
```
Comment (lightmark) [inline execution/executor/src/components/chunk_output.rs:55]: why?

Comment (grao1991) [reply execution/executor/src/components/chunk_output.rs:55]: to reuse prime_cache_by_write_set func

Comment (lightmark) [reply execution/executor/src/components/chunk_output.rs:55]: why do we need clone and call `into`. I think you can just change the `HashSet` to `Vec`

Comment (lightmark) [inline execution/executor/src/components/chunk_output.rs:55]:
```suggestion
let write_sets = transaction_outputs
            .iter()
            .map(|o| o.write_set())
            .collect::<Vec<_>>();
```

File: storage/storage-interface/src/cached_state_view.rs [hunk 3]
```diff
@@ -104,13 +111,20 @@ impl CachedStateView {
     }
 
     pub fn prime_cache_by_write_set(&self, write_sets: &[WriteSet]) -> Result<()> {
```
Comment (lightmark) [inline storage/storage-interface/src/cached_state_view.rs:113]:
```
pub fn prime_cache_by_write_set<'a, T: IntoIterator<Item = &'a WriteSet> + Send>(&self, write_sets: T) -> Result<()> {
        IO_POOL.scope(|s| {
            write_sets
                .into_iter()
```

**Review by lightmark — COMMENTED** [review]
we can avoid copy by...

File: storage/storage-interface/src/cached_state_view.rs (hunk 3 above)
Comment (grao1991) [reply storage/storage-interface/src/cached_state_view.rs]: Done

File: execution/executor/src/components/chunk_output.rs (hunk 2 above)
Comment (grao1991) [reply execution/executor/src/components/chunk_output.rs]: Done

Comment (grao1991) [reply execution/executor/src/components/chunk_output.rs]: Done

File: storage/aptosdb/src/state_store/mod.rs [hunk 4]
```diff
@@ -518,7 +531,36 @@ impl StateStore {
...
+                    s.spawn(move |_| {
+                        let _timer = OTHER_TIMERS_SECONDS
+                            .with_label_values(&["put_stats_and_indices__get_state_value"])
+                            .start_timer();
+                        let version_and_value = self
+                            .state_db
+                            .get_state_value_with_version_by_version(key, base_version)
+                            .expect("Must succeed.");
+                        if let Some((version, value)) = version_and_value {
+                            cache.lock().insert(key.clone(), (version, Some(value)));
```
Comment (sitalkedia) [inline storage/aptosdb/src/state_store/mod.rs:556]: Use a concurrent hash map here instead of locking?

Comment (sitalkedia) [reply storage/aptosdb/src/state_store/mod.rs:556]: https://github.com/xacrimon/dashmap- This is something we seem to be using in other places for lock free hashmap.

Comment (msmouse) [reply storage/aptosdb/src/state_store/mod.rs:556]: Dashmap is slower at least when we tried to make SchemaBatch concurrent -- not a bottleneck, I won't even bother to try again here.

Comment (grao1991) [reply storage/aptosdb/src/state_store/mod.rs:556]: I remember @msmouse said the dashmap is not that good, worse than his sharded hash map..

Given this part is not the bottleneck after this change, I'd prefer to leave it as is, and do some investigation first.

Comment (msmouse) [reply storage/aptosdb/src/state_store/mod.rs:556]: @grao1991  in the case of SchemaBatch it's slower than a plain Mutex<HashMap>, not even sharded.

**Review by lightmark — APPROVED** [review]

**Pull request merged**